x/perp/v2/keeper/dnr.go

```diff
@@ -68,18 +97,78 @@ func (k Keeper) gcUserVolume(ctx sdk.Context, user sdk.AccAddress, currentEpoch
 // Returns zero if the user has no volume for the last epoch.
-func (k Keeper) GetUserVolumeLastEpoch(ctx sdk.Context, user sdk.AccAddress) math.Int {
-	currentEpoch, err := k.DnREpoch.Get(ctx)
-	if err != nil {
-		// a DnR epoch should always exist, otherwise it means the chain was not initialized properly.
-		panic(err)
-	}
+func (k Keeper) GetTraderVolumeLastEpoch(ctx sdk.Context, currentEpoch uint64, user sdk.AccAddress) math.Int {
 	// if it's the first epoch, we do not have any user volume.
 	if currentEpoch == 0 {
 		return math.ZeroInt()
 	}
 	// return the user's volume for the last epoch, or zero.
 	return k.TraderVolumes.GetOr(ctx, collections.Join(user, currentEpoch-1), math.ZeroInt())
 }
+
+// GetTraderDiscount will check if the trader has a custom discount for the given volume.
+// If it does not have a custom discount, it will return the global discount for the given volume.
+// The discount is the nearest left entry of the trader volume.
+func (k Keeper) GetTraderDiscount(ctx sdk.Context, trader sdk.AccAddress, volume math.Int) (math.LegacyDec, bool) {
+	// we try to see if the trader has a custom discount.
+	customDiscountRng := collections.PairRange[sdk.AccAddress, math.Int]{}.
+		Prefix(trader).
+		EndInclusive(volume).
+		Descending()
+
+	customDiscount := k.TraderDiscounts.Iterate(ctx, customDiscountRng)
+	defer customDiscount.Close()
+
+	if customDiscount.Valid() {
+		return customDiscount.Value(), true
+	}
+
+	// if it does not have a custom discount we try with global ones
+	globalDiscountRng := collections.Range[math.Int]{}.
+		EndInclusive(volume).
+		Descending()
+
+	globalDiscounts := k.GlobalDiscounts.Iterate(ctx, globalDiscountRng)
+	defer globalDiscounts.Close()
+
+	if globalDiscounts.Valid() {
+		return globalDiscounts.Value(), true
+	}
+	return math.LegacyZeroDec(), false
+}
+
+// applyDiscountAndRebate applies the discount and rebate to the given exchange fee ratio.
+// It updates the current epoch trader volume.
+// It returns the new exchange fee ratio.
+func (k Keeper) applyDiscountAndRebate(
+	ctx sdk.Context,
+	_ asset.Pair,
+	trader sdk.AccAddress,
+	positionNotional math.LegacyDec,
+	exchangeFeeRatio sdk.Dec,
+) (exFeeRatio sdk.Dec, err error) {
+	// update user volume
+	dnrEpoch, err := k.DnREpoch.Get(ctx)
+	if err != nil {
+		return
+	}
+	k.IncreaseTraderVolume(ctx, dnrEpoch, trader, positionNotional.Abs().TruncateInt())
+
+	// get past epoch volume
+	pastVolume := k.GetTraderVolumeLastEpoch(ctx, dnrEpoch, trader)
+	// if the trader has no volume for the last epoch, we return the provided fee ratios.
+	if pastVolume.IsZero() {
+		return exchangeFeeRatio, nil
+	}
+
+	// try to apply discount
+	discount, hasDiscount := k.GetTraderDiscount(ctx, trader, pastVolume)
+	// if the trader does not have any discount, we return the provided fee ratios.
+	if !hasDiscount {
+		return exchangeFeeRatio, nil
+	}
+	// return discounted fee ratios
+	return discount, nil
```

Comment:
I would rename `discount` to `traderDiscountedFee` since the term `discount` is usually applied to an existing fee, whereas the `TraderDiscount` map holds the actual fee ratio.

Reply:
He has the expressive variable name for free in the function signature
```go
func (k Keeper) applyDiscountAndRebate(
	ctx sdk.Context,
	_ asset.Pair,
	trader sdk.AccAddress,
	positionNotional math.LegacyDec,
	exchangeFeeRatio sdk.Dec,
) (exFeeRatio sdk.Dec, err error)
```
`exFeeRatio` <-> `discount`

Reply:
done!